tests/integration/WC_Facebookcommerce_Integration_Test.php

```diff
@@ -771,8 +773,396 @@ public function test_is_product_sync_enabled_filter() {
+	}
+
+
+	/** @see \WC_Facebookcommerce_Integration::on_product_delete() */
+	public function test_on_product_delete_with_product_variation() {
+
+		$variation = $this->tester->get_product_variation( [ 'status' => 'publish' ] );
+
+		$integration = $this->make( \WC_Facebookcommerce_Integration::class, [
+			'delete_product_item' => Expected::never(),
+			'get_product_fbid'    => Expected::never(),
+		] );
+
+		$integration->on_product_delete( $variation->get_id() );
+
+		$this->tester->assertProductsAreScheduledForDelete( [ $variation->get_id() ] );
+	}
+
+
+	/** @see \WC_Facebookcommerce_Integration::on_product_delete() */
+	public function test_on_product_delete_with_invalid_product_id() {
+
+		$this->tester->setPropertyValue( facebook_for_woocommerce()->get_products_sync_handler(), 'requests', [] );
+
+		$integration = $this->make( \WC_Facebookcommerce_Integration::class, [
+			'delete_product_item' => Expected::never(),
+			'get_product_fbid'    => Expected::never(),
+		] );
+
+		$integration->on_product_delete( 0 );
+
+		$this->tester->assertProductsAreNotScheduledForDelete();
+	}
+
+
+	/**
+	 * @see \WC_Facebookcommerce_Integration::fb_change_product_published_status()
+	 *
+	 * @param string $new_status
+	 * @param string $old_status
+	 * @param \WP_post $post
+	 *
+	 * @dataProvider provider_fb_change_product_published_status
+	 */
+	public function test_fb_change_product_published_status( $new_status, $old_status ) {
+
+		$product = $this->tester->get_product( [ 'status' => $new_status ] );
+
+		$integration = $this->make( \WC_Facebookcommerce_Integration::class, [
+			'update_fb_visibility' => Expected::once(),
+		] );
+
+		$integration->fb_change_product_published_status( $new_status, $old_status, get_post( $product->get_id() ) );
+	}
+
+
+	/** @see test_fb_change_product_published_status() */
+	public function provider_fb_change_product_published_status() {
+
+		return [
+			'publish'   => [ 'publish', 'trash' ],
+			'unpublish' => [ 'trash', 'publish' ],
+		];
+	}
+
+
+	/** @see \WC_Facebookcommerce_Integration::fb_change_product_published_status() */
+	public function test_fb_change_product_published_status_with_invalid_product() {
+
+		$integration = $this->make( \WC_Facebookcommerce_Integration::class, [
+			'update_fb_visibility' => Expected::never(),
+		] );
+
+		$integration->fb_change_product_published_status( null, null, (object) [ 'ID' => 0 ] );
+	}
+
+
+	/** @see \WC_Facebookcommerce_Integration::on_product_publish() */
+	public function test_on_product_publish_with_simple_product() {
+
+		$this->check_on_product_publish( $this->tester->get_product( [ 'status' => 'publish' ] ) );
+	}
+
+
+	/**
+	 * Tests that on_product_publish() delegates to the appropriate handler.
+	 *
+	 * @see \WC_Facebookcommerce_Integration::on_product_publish()
+	 *
+	 * @param \WC_Product $product product object
+	 */
+	private function check_on_product_publish( \WC_Product $product ) {
```

Comment:
This one feels a bit weird to me. I expected this method to be a test method and `test_on_product_publish_with_simple_product` and `test_on_product_publish_with_variable_product ` to be set in a provider.

Reply:
@deivamagalhaes I started adding these helper methods because I was unable to access `$this->tester` in the provider. Although I'm not sure why and maybe there is a way to make it available.

Reply:
@deivamagalhaes I took another look at this one and found why the tester is not available in the data provider.

The tester is injected in the [`_setUp()` method](https://github.com/Codeception/Codeception/blob/4691ecac31992aba50a7c1c58935a76be29ae8c2/src/Codeception/Test/Unit.php#L50-L53), which is called from PHPUnit's [`TestCase::setup()`](https://github.com/Codeception/phpunit-wrapper/blob/e5edaad5583e74165c59d710a8e49a462b1e4a7a/src/TestCase.php#L12). The problem is that those methods are executed [after the data providers are called](https://phpunit.de/manual/3.7/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers):

>All data providers are executed before both the call to the setUpBeforeClass static method and the first call to the setUp method. Because of that you can't access any variables you create there from within a data provider. This is required in order for PHPUnit to be able to compute the total number of tests.

I tried passing just the product properties, but since I use `get_product()` in one case and `get_variable_product()` in the other, I'm forced to add a conditional in `check/test_on_product_publish()` that doesn't seem to help with making the implementation more clear.

What do you think?

Reply:
I'm fine with that. Thanks for clarifying.